plugins/SpectrumAnalyzer/SaProcessor.cpp

```diff
@@ -0,0 +1,556 @@
+	}
+}
+
+
+float SaProcessor::getFreqRangeMax()
+{
+	switch (m_controls->m_freqRangeModel.value())
+	{
+		case FRANGE_AUDIBLE: return FRANGE_AUDIBLE_END;
+		case FRANGE_BASS: return FRANGE_BASS_END;
+		case FRANGE_MIDS: return FRANGE_MIDS_END;
+		case FRANGE_HIGH: return FRANGE_HIGH_END;
+		default:
+		case FRANGE_FULL: return getSampleRate() / 2;
+	}
+}
+
+
+// Map frequency to pixel x position on a display of given width.
+float SaProcessor::freqToXPixel(float freq, int width)
+{
+	if (m_controls->m_logXModel.value())
+	{
+		if (freq <= 1) {return 0;}
+		float min = log10(getFreqRangeMin());
+		float range = log10(getFreqRangeMax()) - min;
+		return (log10(freq) - min) / range * width;
+	}
+	else
+	{
+		float min = getFreqRangeMin();
+		float range = getFreqRangeMax() - min;
+		return (freq - min) / range * width;
+	}
+}
+
+
+// Map pixel x position on display of given width back to frequency.
+float SaProcessor::xPixelToFreq(float x, int width)
+{
+	if (m_controls->m_logXModel.value())
+	{
+		float min = log10(getFreqRangeMin());
+		float max = log10(getFreqRangeMax());
+		float range = max - min;
+		return pow(10, min + x / width * range);
+	}
+	else
+	{
+		float min = getFreqRangeMin();
+		float range = getFreqRangeMax() - min;
+		return min + x / width * range;
+	}
+}
+
+
+// --------------------------------------
+// Amplitude conversion helpers
+//
+float SaProcessor::getAmpRangeMin(bool linear)
+{
+	// return very low limit to make sure zero gets included at linear grid
+	if (linear) {return -900;}
+	switch (m_controls->m_ampRangeModel.value())
+	{
+		case ARANGE_EXTENDED: return ARANGE_EXTENDED_START;
+		case ARANGE_AUDIBLE: return ARANGE_AUDIBLE_START;
+		case ARANGE_NOISE: return ARANGE_NOISE_START;
+		default:
+		case ARANGE_DEFAULT: return ARANGE_DEFAULT_START;
+	}
+}
+
+
+float SaProcessor::getAmpRangeMax()
+{
+	switch (m_controls->m_ampRangeModel.value())
+	{
+		case ARANGE_EXTENDED: return ARANGE_EXTENDED_END;
+		case ARANGE_AUDIBLE: return ARANGE_AUDIBLE_END;
+		case ARANGE_NOISE: return ARANGE_NOISE_END;
+		default:
+		case ARANGE_DEFAULT: return ARANGE_DEFAULT_END;
+	}
+}
+
+
+// Map amplitude to pixel y position on a display of given height.
+float SaProcessor::ampToYPixel(float amplitude, int height)
+{
+	if (m_controls->m_logYModel.value())
+	{
+		if (10 * log10(amplitude) < getAmpRangeMin())
+		{
+			return height;
+		}
+		else
+		{
+			float max = getAmpRangeMax();
+			float range = getAmpRangeMin() - max;
+			return (10 * log10(amplitude) - max) / range * height;
+		}
+	}
+	else
+	{
+		float max = pow(10, getAmpRangeMax() / 10);
+		float range = pow(10, getAmpRangeMin() / 10) - max;
+		return (amplitude - max) / range * height;
+	}
+}
+
+
+// Map pixel y position on display of given height back to amplitude.
+float SaProcessor::yPixelToAmp(float y, int height)
+{
+	if (m_controls->m_logYModel.value())
+	{
+		float max = getAmpRangeMax();
+		float range = getAmpRangeMin() - max;
+		return max + range * (y / height);
```

Comment:
It's not missing (otherwise the cursor measurement wouldn't work), but I probably made some assumptions about using the functions that are not documented o_O
Specifically, I think `ampToYPixel()` always gets linear amplitude parameter (and displays it in a linear or logarithmic way based on config). The `yPixelToAmp()` does the opposite, but in log mode it returns the amplitude in dB, because the user most likely wants dB reading on a dB scale. So the naming may suggest the functions are inverse of one another, but they are not.

Reply:
I updated the comments and added a note specifically stating the second function behaves differently than may be assumed from the first one. Hopefully it is clearer now.

Reply:
OK, I'll hit resolve for this one then.